tokio-quiche/src/quic/io/worker.rs

```diff
@@ -475,26 +498,56 @@ where
         let mut send_buf = &mut send_buf[self.write_state.bytes_written..];
         if send_buf.len() > segment_size.unwrap_or(usize::MAX) {
             // Never let the buffer be longer than segment size, for GSO to
-            // function properly
+            // function properly.
             send_buf = &mut send_buf[..segment_size.unwrap_or(usize::MAX)];
         }
 
-        match qconn.send(send_buf) {
+        // On the first call to `select_path()` a path will be chosen based on
+        // the local address the connection initially landed on. Once a path is
+        // selected following calls to `select_path()` will return it, until it
+        // is reset at the start of the next write cycle.
+        //
+        // The path is then passed to `send_on_path()` which will only generate
+        // packets meant for that path, this way a single GSO buffer will only
+        // contain packets that belong to the same network path, which is
+        // required because the from/to addresses for each `sendmsg()` call
+        // apply to the whole GSO buffer.
+        let (from, to) = self.select_path(qconn).unzip();
+
+        match qconn.send_on_path(send_buf, from, to) {
             Ok((packet_size, info)) => {
                 let _ = send_info.get_or_insert(info);
 
                 self.write_state.bytes_written += packet_size;
                 self.write_state.num_pkts += 1;
-                self.write_state.send_from =
-                    send_info.as_ref().map(|info| info.from);
+
+                let from = send_info.as_ref().map(|info| info.from);
+                let to = send_info.as_ref().map(|info| info.to);
+
+                self.write_state.selected_path = from.zip(to);
+
+                self.write_state.has_pending_data = true;
 
                 Ok(packet_size)
             },
+
             Err(QuicheError::Done) => {
-                // Flush to network and yield when there are no
-                // more packets to write.
+                // Flush the current buffer to network. If no other path needs
+                // to be flushed to the network also yield the work loop task.
+                //
+                // Otherwise the write loop will start again and the next path
+                // will be selected.
+                let has_pending_paths = self.write_state.pending_paths.len() > 0;
+                let has_flushed_paths = self.write_state.has_flushed_paths;
```

Comment:
AFAICT we never set `self.write_state.has_flushed_paths` to `true`, I think we'd want to set it in `flush_buffer_to_socket`?

More generally, can you elaborate on why do we need to continue writing if we've already flushed data? I must be missing something, as it seems to me that we could just loop so long as `has_pending_paths` is true if `send_on_path` returns `Error::Done`